plonky2/src/gadgets/lookup.rs

```diff
@@ -89,25 +89,53 @@ impl<F: RichField + Extendable<D>, const D: usize> CircuitBuilder<F, D> {
+                 `add_gate` call.
+                */
+                // lookup_iter will iterate over the lookups that can be placed in fully utilized
+                // gates, splitting them in chunks that can be placed in the same `LookupGate`
+                let lookup_iter = lookups.chunks_exact(num_slots);
+                // `last_chunk` will contain the remainder of lookups, which cannot fill all the
+                // slots of a `LookupGate`; this last chunk will be processed by incrementally
+                // filling slots, to avoid that the `LookupGenerator` is run on unused slots
+                let last_chunk = lookup_iter.remainder();
+                // handle chunks that can fill all the slots of a `LookupGate`
+                lookup_iter.for_each(|chunk| {
+                    let row = self.add_gate(gate.clone(), vec![]);
+                    for (i, (looking_in, looking_out)) in chunk.iter().enumerate() {
+                        let gate_in = Target::wire(row, LookupGate::wire_ith_looking_inp(i));
+                        let gate_out = Target::wire(row, LookupGate::wire_ith_looking_out(i));
+                        self.connect(gate_in, *looking_in);
+                        self.connect(gate_out, *looking_out);
+                    }
+                });
+                // deal with the last chunk
+                for (_, (looking_in, looking_out)) in last_chunk.iter().enumerate() {
                     let (gate, i) =
-                        self.find_slot(gate, &[F::from_canonical_usize(lut_index)], &[]);
+                        self.find_slot(gate.clone(), &[F::from_canonical_usize(lut_index)], &[]);
                     let gate_in = Target::wire(gate, LookupGate::wire_ith_looking_inp(i));
                     let gate_out = Target::wire(gate, LookupGate::wire_ith_looking_out(i));
-                    self.connect(gate_in, looking_in);
-                    self.connect(gate_out, looking_out);
+                    self.connect(gate_in, *looking_in);
+                    self.connect(gate_out, *looking_out);
                 }
 
                 // Create LUT gates. Nothing is connected to them.
                 let last_lut_gate = self.num_gates();
                 let num_lut_entries = LookupTableGate::num_slots(&self.config);
                 let num_lut_rows = (self.get_luts_idx_length(lut_index) - 1) / num_lut_entries + 1;
-                let num_lut_cells = num_lut_entries * num_lut_rows;
-                for _ in 0..num_lut_cells {
+                // Also for `LookupTableGate` we replace the `num_lut_entries` calls to `find_slot`
+                // with a single call to `add_gate`; note that in this case there is no need to
+                // separately handle the last chunk of LUT entries that cannot fill all the slots of
+                // a `LookupTableGate`, as the generator already handles empty slots
```

Comment:
nit: This comment makes sense in the context of this PR, but not in general.

Reply:
I just meant that the comment made sense when looking at the code you replaced, but not so much to an external reader not looking at the history of the file.
But the new comment looks good 👍